### PR TITLE
closes #3497 - Ensures that an endpoint property can be provided to Auth#configure

### LINKED ISSUE #3497: [FEATURE] Allow for @aws-amplify/auth configure to set an endpoint
**Is your feature request related to a problem? Please describe.**

AWS cognito exposes a URL  that we are meant to use in order to consume the service. e,g `https://cognito-idp.eu-west-1.amazonaws.com/`

Said cognito URL and params to make a request are out in the public to see when `amplify` is being used as a client-side library.

That leaves a room open for attackers to use and abuse the endpoint.

As a solution, a development team can set a forward proxy service to be behind their own CDN in order to conceal the fact Cognito is being used and programmatically implement strategies like rate-limiting, WAF IP blacklisting, etc.

However, the SDK does not allow for an end user to set up an endpoint.

**Describe the solution you'd like**

I would like the `Auth.configure();` method to be able to take in the [`endpoint`](https://github.com/aws-amplify/amplify-js/blob/master/packages/amazon-cognito-identity-js/src/CognitoUserPool.js#L36) param. e,g:

```js
      Auth.configure({
       endpoint: config('PROXIED_ENDPOINT'),
        region: config('COGNITO_REGION'),
        userPoolId: config('COGNITO_USER_POOL'),
        userPoolWebClientId: config('COGNITO_CLIENT_ID'),
      })
```

This would:

- Give the end users a lot more control over what could potentially be happening and going through Cognito

- Make testing a lot easier.

### PR DESCRIPTION
**Pull request by mariogintili**
*Issue #, if available:*

#3497

*Description of changes:*

- Document `data.endpoint` on `CognitoUserPool`
- Extends `Auth.ts` to forward an `endpoint` property to `CognitoUserPool`
- Extends type `AuthOptions ` to allow for an endpoint property as a string or not
- Adds test coverage to ensure the above is functional

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

**NOTE** I did not bump the package since I'm unsure as to what type of version update this type of change corresponds to. Suggestions welcome

### CODE REVIEW

File: packages/amazon-cognito-identity-js/src/CognitoUserPool.js [hunk 1]
```diff
@@ -33,13 +34,19 @@ export default class CognitoUserPool {
...
+    const { UserPoolId, ClientId, AdvancedSecurityDataCollectionFlag, endpoint } = data || {};
     if (!UserPoolId || !ClientId) {
       throw new Error('Both UserPoolId and ClientId are required.');
     }
     if (!/^[\w-]+_.+$/.test(UserPoolId)) {
       throw new Error('Invalid UserPoolId format.');
     }
+
+    // eslint-disable-next-line max-len
+    if (endpoint && /^(http|https):\/\/.+/.test(endpoint)) {
```
Comment (ricoli) [inline packages/amazon-cognito-identity-js/src/CognitoUserPool.js]: shouldn't the second condition be negated, as `test` will return `true` when there's a match and the outcome is to throw an error?

**Comment by razor-x** [comment]
I would really like to see this merged, can a maintainer please note what is required to merge? I can help push this through if more work is needed.

Thanks

**Comment by razor-x** [comment]
Does anyone know if the user pool id would still be required if using an endpoint that points to the user pool? 

If not then the conditional check could be updated to allow no pool id if an endpoint is given. This would simplify configuration and not expose the pool id.

**Comment by mariogintili** [comment]
@razor-x I agree with not exposing a pool-id. That way you could have a proxy as an endpoint that just forwards requests + the pool ID. That way your pool ID is kept as a secret server side 😄

**Comment by justcompile** [comment]
Would really like to see this merged as we're facing the same issue.

**Comment by mariogintili** [comment]
I've abandoned this PR as I was developing this stuff for a client who doesn't really deserve it ⭐️ . Happy to have someone else take over

**Comment by rebolyte** [comment]
@mariogintili I'm facing this same issue and would be interested in taking over this PR. If you're willing, would you rather give write access to your fork, transfer ownership, or have a new PR started?

**Comment by mariogintili** [comment]
@rebolyte please feel free to take over the PR in any shape or form 😄

**Comment by rebolyte** [comment]
@mariogintili Seems easiest if you add me as a collaborator to your fork? Then we can keep everything in the same place.

**Comment by rebolyte** [comment]
Hey @jordanranz @sammartinez (or other contributors)! I've resolved the conflicts here, do you see anything else that needs attention before merging this in? Would be a huge help.

**Comment by razor-x** [comment]
I still have a few questions about this. mainly with the goal of making `userPoolId` not needed as a param if `endpoint` is passed. https://github.com/aws-amplify/amplify-js/blob/73cdf815c617dae89e0f230924392fb021fa853a/packages/amazon-cognito-identity-js/src/CognitoUserPool.js#L43-L45

I see that `userPoolId` is all over `Auth.js` but it seems like `amazon-cognito-identity-js` is much simpler and might have a way forward for `endpoint` OR `userPoolId`.

1. This line still depends on a  `userPoolId`. The `region` should be an optional parameter and string-based detection on the  `userPoolId` the fallback. https://github.com/aws-amplify/amplify-js/blob/73cdf815c617dae89e0f230924392fb021fa853a/packages/amazon-cognito-identity-js/src/CognitoUserPool.js#L56
2. If endpoint is given, then why is region still needed? https://github.com/aws-amplify/amplify-js/blob/73cdf815c617dae89e0f230924392fb021fa853a/packages/amazon-cognito-identity-js/src/CognitoUserPool.js#L61
3. This is the only other spot I see that depends on it, do we know if we can work around it? https://github.com/aws-amplify/amplify-js/blob/73cdf815c617dae89e0f230924392fb021fa853a/packages/amazon-cognito-identity-js/src/CognitoUserPool.js#L173-L177

**Comment by rebolyte** [comment]
@haverchuck 👋 Any updates here?

**Comment by whummer** [comment]
Thanks a lot for working on this @mariogintili @ricoli @razor-x @rebolyte ! This is indeed a much needed feature, to simplify testing of applications locally. It's a shame that these types of changes are not accepted by the maintainers.

Meanwhile, we have created a library `amplify-js-local@0.1.0` that provides an `amplifylocal` CLI to use Amplify and deploy applications entirely against local APIs (LocalStack). This first version we created is only a starting point - it applies simple runtime patching to configure custom endpoints and test credentials.

Perhaps it can help support your use case well - please feel free to raise issues or pull requests [in this repo](https://github.com/localstack/amplify-js-local). We'll be happy to incorporate any changes that may simplify using Amplify with custom endpoints - looking forward to getting your feedback! Cheers

**Comment by sammartinez** [comment]
@rebolyte or @mariogintili Can you resolve the conflicts on here?

**Comment by slaymance** [comment]
I resolved conflicts the conflicts for this PR and it should be ready to merge pending final review. I created a separate issue ([Issue 8244](https://github.com/aws-amplify/amplify-js/issues/8244) to track the suggestion by @razor-x of allowing `userPoolId` to be optional in favor of an exposed `endpoint` value. This will require some additional discovery and would be best addressed in a separate PR.

**Comment by slaymance** [comment]
Since what this PR was trying to accomplish was mostly addressed by [PR 6124](https://github.com/aws-amplify/amplify-js/pull/6124), I'm closing it and tracking the additional test coverage for the `endpoint` configuration option here: [Issue 8244](https://github.com/aws-amplify/amplify-js/issues/8244)

**Pull request closed without merging**